tests/test_conn.py

```diff
@@ -0,0 +1,161 @@
+
+    cmd = shell.build_command_line("echo hello world", cwd=None, env={"HELLO": "WORLD", "JOHN": "DOE"})
+    expected = textwrap.dedent(
+        """
+        export HELLO=WORLD
+        export JOHN=DOE
+
+        echo hello world
+        """
+    ).strip()
+    assert cmd == f"{shell.shell_command} '{expected}'"
+
+
+def test_conn__shell_bash__cwd_env():
+    shell = Bash()
+
+    assert shell.name == "bash"
+    assert shell.shell_command == "/usr/bin/bash -c"
+
+    cmd = shell.build_command_line("echo hello world", cwd="/home/test", env={"HELLO": "WORLD", "JOHN": "DOE"})
+    expected = textwrap.dedent(
+        """
+        export HELLO=WORLD
+        export JOHN=DOE
+        cd /home/test
+
+        echo hello world
+        """
+    ).strip()
+    assert cmd == f"{shell.shell_command} '{expected}'"
+
+
+@pytest.mark.parametrize(
+    "input, expected",
+    [
+        ("Write-Output hello world", "Write-Output hello world"),
+        ('Write-Output "hello world"', 'Write-Output \\"hello world\\"'),
+        ("Write-Output 'hello world'", "Write-Output ''hello world''"),
+    ],
+    ids=[
+        "no-quotes",
+        "double-quotes",
+        "single-quotes",
+    ],
+)
+def test_conn__shell_powershell__script(input: str, expected: str):
+    shell = Powershell()
+
+    assert shell.name == "powershell"
+    assert shell.shell_command == "powershell -NonInteractive -Command"
+
+    cmd = shell.build_command_line(input, cwd=None, env={})
+    assert cmd == f"{shell.shell_command} '{expected}'"
+
+
+def test_conn__shell_powershell__cwd():
+    shell = Powershell()
+
+    assert shell.name == "powershell"
+    assert shell.shell_command == "powershell -NonInteractive -Command"
+
+    cmd = shell.build_command_line("Write-Output hello world", cwd="/home/test", env={})
+    expected = textwrap.dedent(
+        """
+        cd /home/test
+
+        Write-Output hello world
+        """
+    ).strip()
+    assert cmd == f"{shell.shell_command} '{expected}'"
+
+
+def test_conn__shell_powershell__env():
+    shell = Powershell()
+
+    assert shell.name == "powershell"
+    assert shell.shell_command == "powershell -NonInteractive -Command"
+
+    cmd = shell.build_command_line("Write-Output hello world", cwd=None, env={"HELLO": "WORLD", "JOHN": "DOE"})
+    expected = textwrap.dedent(
+        """
+        $Env:HELLO = WORLD
+        $Env:JOHN = DOE
+
+        Write-Output hello world
+        """
+    ).strip()
+    assert cmd == f"{shell.shell_command} '{expected}'"
+
+
+def test_conn__shell_powershell__cwd_env():
+    shell = Powershell()
+
+    assert shell.name == "powershell"
+    assert shell.shell_command == "powershell -NonInteractive -Command"
+
+    cmd = shell.build_command_line("echo hello world", cwd="/home/test", env={"HELLO": "WORLD", "JOHN": "DOE"})
+    expected = textwrap.dedent(
+        """
+        $Env:HELLO = WORLD
+        $Env:JOHN = DOE
+        cd /home/test
+
+        echo hello world
```

Comment:
better to use powershell commands like `Set-Location` and `Write-Output` here?

Reply:
It does not matter since the code is not really executed, but cd and echo are valid commands. I think they are actually aliases.

Reply:
Ok no problem.